app.js

```diff
@@ -73,89 +72,96 @@ function _init() {
     && config.server.ssl_key && config.server.ssl_port;
 }
 
-function refineRemotes(myRemotes) {
-  var newRemotes = {};
-  var newRemoteCommands = null;
-  var remote = null;
-
-  function isBlacklistExisting(remoteName) {
-    return config.blacklists && config.blacklists[remoteName];
+function overrideConfigurationForDebugOrDevelopment() {
+  var lircTest;
+  if (process.env.npm_package_config_test_env) {
+    console.log('we are in test mode!');
+    lircTest = require('./test/lib/lirc');
+    lircTest.replaceLircByMock();
+    gpio.setGpioLibrary(require('./lib/gpio-la-mock'));
+    config = require('./test/fixtures/config.json');
+    hasServerPortConfig = false;
+    hasSSLConfig = false;
+  } else {
+    gpio.setGpioLibrary(require('./lib/gpio-la-gpio'));
   }
+}
 
-  function getCommandsForRemote(remoteName) {
-    var remoteCommands = myRemotes[remoteName];
-    var blacklist = null;
-
-    if (isBlacklistExisting(remoteName)) {
-      blacklist = config.blacklists[remoteName];
+function initializeModules(done) {
+  function initializeMacros() {
+    var currentStates = macros.getCurrentStates();
+    macros.resetConfiguration();
 
-      remoteCommands = remoteCommands.filter(function (command) {
-        return blacklist.indexOf(command) < 0;
-      });
+    if (config.gpios) {
+      macros.registerDevice(gpio);
     }
-
-    return remoteCommands;
+    macros.registerDevice(lirc);
+    macros.init(config.macros, currentStates);
   }
 
-  for (remote in myRemotes) {
-    newRemoteCommands = getCommandsForRemote(remote);
-    newRemotes[remote] = newRemoteCommands;
-  }
+  lirc.init(config, function () {
+    if (config.gpios) {
+      gpio.init(config.gpios);
```

Comment:
Is there a reason why `gpio` initialization is done within `lirc.init` callback?

Reply:
Yeah. Lirc consumes the whole configuration, including macros. This implies a lack of separation of concern, which leads to cross dependencies while initalizing both. 

I'd prefer a initialization like iterate on configured modules, initialize them, initialize macros. This would imply a restructuring of the configuration.

Could we take it for now? I don't want to do two refactorings in one pull request. Or would you mind to suggest another aproach for perfoming initialisation?